src/python/pants/base/specs_parser.py

```diff
@@ -84,8 +88,21 @@ def parse_spec(self, spec: str) -> Union[AddressSpec, FilesystemSpec]:
         spec_path = self._normalize_spec_path(spec)
         if Path(self._root_dir, spec_path).is_file():
             return FilesystemLiteralSpec(spec_path)
-        # Else we apply address shorthand, i.e. `src/python/pants/util` -> `src/python/pants/util:util`
-        # TODO: Figure out what this should look like if (once?) filesystem specs allow directories.
-        # Should this shorthand be removed so that directories may be unambiguously resolved to a
-        # FilesystemSpec? If so, do we still allow the shorthand in BUILD files?
-        return SingleAddress(directory=spec_path, name=PurePath(spec).name)
+        # Else we apply address shorthand, i.e. `src/python/pants/util` ->
+        # `src/python/pants/util:util`
+        return AddressLiteralSpec(path_component=spec_path, target_component=PurePath(spec).name)
```

Comment:
It might make sense to invert these two cases, and do roughly:
```
if Path(self._root_dir, spec_path).is_dir():
    return AddressLiteralSpec(path_component=spec_path, target_component=PurePath(spec).name)
# Else, either a file or doesn't exist.
return FilesystemLiteralSpec(spec_path)
```

Because it's only if it's a directory that using the basename in an `AddressLiteralSpec` would be correct: otherwise, if the path is either 1) a file, or 2) doesn't exist on disk, our `Owners` logic can kick in via `FilesystemLiteralSpec`.

Reply:
Hm I tried this out and think it ended up being worse. If the path does not exist, the error message becomes "Unmatched globs" from the engine, rather than the more carefully crafted message in `AddressInput -> Address`.

Also, note that we already have a heuristic that file extensions -> `FilesystemLiteralSpec`, so the case of a typo in a file arg will do the right thing already.